File: tarumt-nav-app/expo-app/src/tiled/model/tiledMapLayers.ts
import type {
  TiledMap,
  TiledObjectLayer,
  TiledTileLayer,
} from '../type';

export function assertSupportedMap(map: TiledMap) {
  if (map.type !== 'map') {
    throw new Error(`Unsupported Tiled document type: ${map.type}`);
  }
  if (map.orientation !== 'orthogonal') {
    throw new Error(`Unsupported Tiled orientation: ${map.orientation}`);
  }
  if (!map.infinite) {
    throw new Error('This app expects the demo infinite map export.');
  }
}

export function getVisibleTileLayers(map: TiledMap) {
  return map.layers.filter(
    (layer): layer is TiledTileLayer =>
      layer.type === 'tilelayer' && layer.visible !== false,
  );
}

export function getObjectLayer(map: TiledMap, name: string) {
  return map.layers.find(
    (layer): layer is TiledObjectLayer =>
      layer.type === 'objectgroup' && layer.name === name,
  );
}

File: tarumt-nav-app/flutter_app/lib/domain/tiled/map/tiled_map_layers.dart
import 'package:indoor_navigation/domain/tiled/tiled_models.dart';

void assertSupportedMap(TiledMap map) {
  if (map.orientation != 'orthogonal') {
    throw UnsupportedError('Unsupported Tiled orientation: ${map.orientation}');
  }
  if (map.infinite != true) {
    throw UnsupportedError('This app expects the demo infinite map export.');
  }
}

List<TiledTileLayer> getVisibleTileLayers(TiledMap map) {
  return map.layers
      .whereType<TiledTileLayer>()
      .where((layer) => layer.visible != false)
      .toList(growable: false);
}

TiledObjectLayer? getObjectLayer(TiledMap map, String name) {
  for (final layer in map.layers) {
    if (layer is TiledObjectLayer && layer.name == name) {
      return layer;
    }
  }
  return null;
}
